website/blog/2024-08-06-langgraph-pyfunc/index.md

```diff
@@ -0,0 +1,223 @@
+
+### What is LangGraph?
+
+LangGraph is a library for building stateful, multi-actor applications with LLMs, used to create agent and multi-agent workflows. Compared to other LLM frameworks, it offers these core benefits: cycles, controllability, and persistence. LangGraph allows you to define flows that involve cycles, essential for most agentic architectures, differentiating it from DAG-based solutions. As a very low-level framework, it provides fine-grained control over both the flow and state of your application, crucial for creating reliable agents. Additionally, LangGraph includes built-in persistence, enabling advanced human-in-the-loop and memory features.
+LangGraph is inspired by Pregel and Apache Beam. The public interface draws inspiration from NetworkX. LangGraph is built by LangChain Inc, the creators of LangChain, but can be used without LangChain.
+
+# The Code
+
+### Setup
+First, we must install the required dependencies. We will use OpenAI for our LLM, but LangChain paired with LangGraph makes it easy to substitute your desired LLM.
+
+```python
+%%capture --no-stderr
+%pip install -U langgraph langsmith mlflow
+%pip install --upgrade typing_extensions
+
+# Used for this tutorial; not a requirement for LangGraph
+%pip install -U langchain_openai
+```
+
+Next, let's get our relevant secrets.
+
+```python
+import os
+import getpass
+
+def _set_env(var: str):
+    if not os.environ.get(var):
+        os.environ[var] = getpass.getpass(f"{var}: ")
+
+_set_env("OPENAI_API_KEY")
+_set_env("LANGSMITH_API_KEY")
+```
+
+### Custom Dependencies
+Before building our custom PyFunc model, let's first review what PyFunc does. When you specify a model to log, MLflow will try to leverage `cloudpickle` to store the serialized model. Along with this model artifact, MLflow will store the signature, which can be passed or inferred from the `model_input` parameter. It will also log inferred model dependencies to help you serve the model in a new environment.
+
+However, as with many other packages, pickling objects is often not supported by LangGraph chains. 
+
+To get around this issue, we will leverage the [code_paths](https://mlflow.org/docs/latest/model/dependencies.html?highlight=code_paths#saving-extra-code-with-an-mlflow-model-manual-declaration) argument to specify a custom dependency. Instead of serializing the entire LangGraph chain, we'll look to reload it from a python file. 
+
+Below, we use the magic `%%writefile` command to create a new file in a jupyter notebook context. If you're running this outside of an interactive notebook, simply create the file below, omitting the `%%writefile graph_chain.py` line.
+
+```python
+%%writefile graph_chain.py
```

Comment:
I would add a comment here for those readers who only read the code blocks

#omit this line if directly creating this file; this command is purely for running within Jupyter